source/fuzz/transformation_access_chain.cpp

```diff
@@ -196,21 +280,149 @@ protobufs::Transformation TransformationAccessChain::ToMessage() const {
 }
 
 std::pair<bool, uint32_t> TransformationAccessChain::GetIndexValue(
-    opt::IRContext* ir_context, uint32_t index_id) const {
+    opt::IRContext* ir_context, uint32_t index_id,
+    uint32_t object_type_id) const {
+  if (!ValidIndexToComposite(ir_context, index_id, object_type_id)) {
+    return {false, 0};
+  }
   auto index_instruction = ir_context->get_def_use_mgr()->GetDef(index_id);
-  if (!index_instruction || !spvOpcodeIsConstant(index_instruction->opcode())) {
-    // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3179) We could
-    //  allow non-constant indices when looking up non-structs, using clamping
-    //  to ensure they are in-bounds.
+
+  uint32_t bound = fuzzerutil::GetBoundForCompositeIndex(
+      *ir_context->get_def_use_mgr()->GetDef(object_type_id), ir_context);
+
+  // The index must be a constant
+  if (!spvOpcodeIsConstant(index_instruction->opcode())) {
+    return {false, 0};
+  }
+
+  // The index must be in bounds.
+  uint32_t value = index_instruction->GetSingleWordInOperand(0);
+
+  if (value >= bound) {
+    return {false, 0};
+  }
+
+  return {true, value};
+}
+
+std::pair<bool, uint32_t> TransformationAccessChain::CreateAndGetClampedIndexId(
+    opt::IRContext* ir_context, uint32_t index_id, uint32_t object_type_id,
```

Comment:
I am not keen on this function, nor on TryToClampInteger.  They are complex because they capture both the *checking* that is necessary in IsApplicable, and the *doing* that should only happen in Apply.

I suggest you approach this as follows:

- In "IsApplicable", instead of calling CreateAndGetClampedIndexId, inline all of the checking code so that "IsApplicable" does all of the checks inline.  Similarly, don't call TryToClampInteger - write the code out inline, even if it gets quite long.

- Do the same for "Apply", but only include the "doing" code.  Remember that "Apply" can assume that "IsApplicable" returned true, so e.g. if you need bool type you can use "MaybeGet" to get it, and assert that the result of this call is non-zero; you don't need to check whether it could be found and return false if it wasn't.

Then see how the code looks.  There may well be a bit more code overall, but I expect it will be more readable since in the "IsApplicable" code there will be no logic for changing the module - and no need for a "add_clamping_instructions" to control whether we want to make changes.  Similarly in "Apply" there will be no checking code (just some sanity-checking assertions perhaps).

You might then decide you want to factor some code out into helper functions for readability - but no need to share the helpers between "IsApplicable" and "Apply" unless you think it really makes sense to do so.

Reply:
Done. The duplication in the code actually turned out to be much less than expected, because of all the checks that only IsApplicable needs to do and the actions that only Apply needs to perform.